### ISSUE TITLE
[RFC] Test Cases Enabling for Accelerators

### ISSUE THREAD

**Issue by EikanWang** [pytorch/pytorch#146898]
### 🚀 The feature, motivation and pitch

# Motivation

PyTorch contains a large and continually growing set of test cases to ensure quality and performance across a variety of features. As more device backends (accelerators) are added, contributors frequently face the challenge of enabling those existing test cases for new devices:

1. Device Decorator
PyTorch test cases rely on device-specific decorators—such as onlyCUDA, dtypesIfCUDA, skipIfXLA, etc.—to skip or include tests for particular backends. Enabling a new device often requires adding or modifying these decorators in many tests for reusing.
This creates an intrusive process since broad modifications are required to integrate a device carefully without breaking other backends.

2. Diverse Device Capabilities
Each hardware vendor’s device, or even different generations of a single vendor’s device, may differ in capabilities—e.g., data type support, distributed training support, etc.
A one-size-fits-all approach for test decorators has proven cumbersome when, for instance, a new backend only partially supports a feature (like torch.compile), or only supports certain data types (e.g., no float64 but supports bfloat16).

Thus, we think it is meaningful to create a mechanism that can flexibly determine at runtime which tests to run, skip, or adapt, based on a device’s specific capabilities. This RFC tries to propose ideas to address these issues.

# Proposed Approach

The core idea of this RFC is to introduce device or “accelerator” abstractions that report device capabilities, and then use these capabilities to dynamically configure test inclusion or parameterization. In general, the idea is as follows:

- Define device abstraction to provide a unified interface for querying device capabilities for test case infrastructure
- Provide a registration mechanism allowing devices to dynamically declare their capabilities at runtime.
- Define test case device capability requirements
  - Run test case automatically if a device meets the device capability requirement
  - Skip test case automatically if a device does NOT meet the device capability requirement
 
## Device Abstraction
Create a unified interface—either via torch.accelerator or a specialized test infrastructure interface (similar to DeviceInterface in Dynamo)—to query device capabilities. For example:
```python
class DeviceInterface:
    def device_type(self) -> str:
        ...
    def supports_distributed(self) -> bool:
        ...
    def supported_dtypes(self) -> Set[torch.dtype]:
        ...
    # etc.
```
Alternatively, PyTorch’s existing `torch.accelerator` can be extended or leveraged to provide similar functionality.
Below is a non-exhaustive list of capabilities a device backend might declare, many of which match existing features tested in PyTorch:

- ATen operation
  - whitelist/blacklist  
  - Deterministic
  - (Other capabilities)
- Supports distributed (True/False)  
- Supports torch.compile (True/False)
- Graph mode (CUDA Graph, XPU Graph, etc.)
- Data types (e.g. support for float64, bfloat16, fp8_e5m2, fp8_e4m3, etc.)
- Tensor types (sparse, quantized, etc.)
- Mixed precision
- Quantization features
- OS constraints or platform constraints (if needed)
- Other specialized capabilities

The exact set of fields is flexible, but the concept is that each device or backend can declare these capabilities at runtime, and the testing framework uses them to filter or adapt test cases appropriately.

## Registration of Device Interface

Allow backends to register the device interface to declare their capabilities at runtime. This ensures that:

- Out-of-tree backends (using PrivateUse1 or similar dispatch keys) can inform the testing framework of their features.
- Each device can selectively advertise what it supports (distributed, certain data types, mixed precision, etc.).

### Require capability

In addition to `onlyCPU`, `onlyCUDA` , `dtypesIfCUDA`, etc., the idea proposes decorators that declare capability requirements:
```python
class requires_capabilities:
    def __init__(self, required_caps):
        self.required_caps = required_caps

    def __call__(self, fn):
        @wraps(fn)
        def wrapper(test_self, *args, **kwargs):
            device_iface = get_device_interface()  # Query current device
            if not device_iface.check_device_capability(self.required_caps):
                raise unittest.SkipTest(
                    f"Skipping because device does not meet {self.required_caps}"
                )
            return fn(test_self, *args, **kwargs)
        return wrapper

@requires_capabilities({"distributed": True})
def test_distributed_functionality():
...
```

So that the test framework can decide the test scope(test cases list) at runtime.

- Runtime Skipping/Running: The test framework checks the device’s capabilities, automatically skipping tests that are not supported.
- Runtime Parameterization: Instead of statically declaring which `dtypes` are tested, a test can specify all relevant `dtypes`, and the framework intersects those with the device’s supported `dtypes` to generate the final test set.

### Intersect Capability

Some test cases may require multiple device capabilities, while a particular device may provide partial of the required capabilities. We expect that the test case still can be instantiated. The data types are a good example.

For tests that use @dtypes(...) or a similar approach:
```python
@dtypes(torch.bfloat16, torch.float32, torch.float64)
def test_div_rounding_nonfinite(self, device, type):
    ...
```

- Suppose the device claims to support {torch.bfloat16, torch.float16, torch.float32}.
- The intersection with {torch.bfloat16, torch.float32, torch.float64} is {torch.bfloat16, torch.float32}.
- The framework thus only instantiates:
  - test_div_rounding_nonfinite_accelerator_bfloat16
  - test_div_rounding_nonfinite_accelerator_float32

# Summary
By introducing a unified device-capability abstraction, dynamic capability registration, and capability-based decorators, we expect the RFC can refine the test suite to handle multiple backends gradually.

-  Minimize intrusion in existing tests when bringing up a new device.
-  Provide robust coverage by automatically running all relevant tests for each device’s advertised capabilities.
-  Reduce maintenance overhead by consolidating device-specific logic into a standardized abstraction.

# Context and Limitations
1. Not a One-Size-Fits-All Replacement
Some tests are inherently device-specific. For example, certain CUDA-specific tests rely on features like `cudaStream_t`, kernels that only exist on CUDA, or tooling limited to NVIDIA hardware. These tests should remain decorated with something like `onlyCUDA`.

2. Gradual Adoption
PyTorch has a large test suite with many specialized decorators. This proposal does not require that all tests migrate at once. Instead, new backends or test modules can adopt this approach incrementally.

3. Complementary to Existing Infrastructure
This RFC does not attempt to eliminate or replace all current decorators, nor does it cover every corner case. Instead, it provides a design that can be expanded to other devices or features over time.

**Comment by malfet** [comment]
I'm trying to make OpInfo more extendable by adding generic `dtypesIf`: https://github.com/pytorch/pytorch/pull/146905

**Comment by EikanWang** [comment]
@chuanqi129 , @hipudding FYI

**Comment by shink** [comment]
We are also working on supporting querying device capabilities, which will benefit all privateuse1 devices in more than just testing!

cc: @FFFrog

**Comment by EikanWang** [comment]
@shink , @FFFrog , does the proposed idea align with your work? Any differences?

**Comment by FFFrog** [comment]
@EikanWang, thank you very much for your mention.

I think this proposal is very good, it is not only beneficial to in-tree backends, but also to out-of-tree backends including those based on PrivateUse1, so if possible, we can jointly promote the implementation and implementation of this RFC.

I would like to add that:
Given that PrivateUse1 is a public key, the backend integrated into PyTorch based on PrivateUse1 may have different capabilities, and these differences in capabilities may affect the support and implementation path of some functions within PyTorch; therefore, we hope to provide an API for PrivateUse1-based devices to obtain some functions of the device itself (similar to get_device_properties, but with differences)

I noticed that this RFC also mentioned this from a testing perspective. I would like to know if @EikanWang has any better suggestions in this regard.

@shink FYI.

**Comment by EikanWang** [comment]
In terms of the backends endorsed by `PrivateUse1`, the RFC intends to provide `Registration of Device Interface` to address the point. Currently, we have defined `DeviceInterface` on the dynamo side and provide APIs to register and get the device interface at runtime - https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/device_interface.py

- `register_interface_for_device` 
- `get_registered_device_interfaces`

The idea can be applied to the test infrastructure as well. If the test infrastructure provides such a similar registration mechanism, the `PriveUse1`-based devices can expose its device capabilities to the test infrastructure and then generate test cases accordingly. @FFFrog

**Comment by ZainRizvi** [comment]
This is a really intriguing proposal, esp with how it allows pytorch tests to scale to run against a wide variety of 3rd party hardware.

Questions on my mind:
- **What is the primary use case**? Is it 3rd party vendors who would run pytorch tests in their own CI system? Or is it for the vendors who are already part of the pytorch/pytorch CI? (clearly both can benefit from it, but I'd like to understand who is primarily being catered to here, since they'll likely become test cases to validate the approach).  
- **Adoption**:  How much of a ramp up will this require for pytorch developers? They'll need to be more aware of the device capabilities their tests need.  How much more does that require from them vs `@onlyCUDA`? Having the capabilities formally listed in code will likely help with that ramp up, esp as that list will grow over time.  Tooling to nudge devs in the direction of using these new annotations instead of the old ones would also help.
- **Capability compatibility**: Are capabilities always implemented in a compatible-enough way across hardware device (with the Ops expected to abstract away any differences)?  E.g. for Graph mode the proposal mentions CUDA Graph and XPU Graph. Would tests need to be aware of which hardware they're running against?
 
cc @janeyx99. Would love to get your take on this as a test-writing dev

**Comment by janeyx99** [comment]
If I'm reading this proposal correctly, its goal is to improve our test infra device specialization by leveraging the accelerator abstraction. I think generally this is a worthy goal to attempt as anything that helps clarify our massive system of tests is a good thing, though I'm curious:
1. @EikanWang Is this proposal for a test infra specific DeviceSpec? Or a more generic Accelerator spec? I am anticipating this to be mostly for test infra.
2. Today, we have some logic to parametrize our tests with devices, which has device specific information (e.g., we will not generate CUDA tests on CPU). This device abstraction should integrate well with https://github.com/pytorch/pytorch/blob/ef6b16ea9d200cba87c8730c99264d6d477ab0ec/torch/testing/_internal/common_device_type.py#L882--what's the plan here?

ccing @albanD on accelerator/device genericism
ccing @jbschlosser for test infra

**Comment by EikanWang** [comment]
> What is the primary use case? Is it 3rd party vendors who would run pytorch tests in their own CI system? Or is it for the vendors who are already part of the pytorch/pytorch CI?

This may benefit 3rd party vendors more than in-tree backends. 

> Adoption: How much of a ramp up will this require for pytorch developers?

It is open now and depends on the detailed design and implementation. I hope this approach does not require too much ramp-up effort.

> Capability compatibility: Are capabilities always implemented in a compatible-enough way across hardware device

Yes. For the given example regarding graph mode, test cases/classes need to claim that they require graph capability.

> Is this proposal for a test infra specific DeviceSpec? Or a more generic Accelerator spec? I am anticipating this to be mostly for test infra.

Yes, it is for test infra.

> This device abstraction should integrate well with https://github.com/pytorch/pytorch/blob/ef6b16ea9d200cba87c8730c99264d6d477ab0ec/torch/testing/_internal/common_device_type.py#L882--what's the plan here?

Right, we intend to integrate device capabilities into `common_device_type.py`, not only in `instantiate_device_type_tests` but also in other classes and functions. We will use `dtypes` as the entry point to demonstrate the potential changes.

**Comment by malfet** [comment]
It would be interesting to have some discussion/thoughts on how this will work with different generation of devices for the same accelerator (Volta vs Ampere vs Hopper on CUDA)

**Comment by EikanWang** [comment]
In terms of different generations of devices of the same accelerator, this RFC proposes to inquire the device properties at runtime to conclude the device capabilities. For instance, the fp8 support differs between different generations of HW. Regarding CUDA device, we can check the compute capability to determine whether or not the current device supports FP8.

The concept code is as follows. In the case of functions like `get_valid_dtypes` or `_check_dtype`, each backend can provide its own implementation. Vendors can either build on existing tooling (such as torch.accelerator) or implement their own approach from scratch. (**NOTE:** The code is conventional.)

- Backend to expose interface to query supported data types
```python
import torch
from torch.testing._internal.common_dtype import all_types_complex_float8_and

all_torch_dtypes = all_types_complex_float8_and(torch.half, torch.bfloat16, torch.bool, torch.uint16, torch.uint32, torch.uint64)

class DeviceInterface:
    def __init__(self, device):
        self.device = device

    def _check_dtype(self, dtype):
        return False

    def supported_dtypes(self) -> Set[torch.dtype]:
        return set([dtype for dtype in all_torch_dtypes if self._check_dtype(dtype)])

class CUDADeviceInterface(DeviceInterface):
    def __init__(self, device):
        super().__init__(device)

    def _check_fp8_dtype(self, dtype):
        major, minor = torch.cuda.get_device_capability()
        return major >= 9 and minor >= 0

    def _check_dtype(self, dtype):
        questionable_fp8_dtypes = [torch.float8_e4m3fn, torch.float8_e5m2fn, torch.float8_e5m2, torch.float8_e5m2fnuz]
        if dtype in questionable_fp8_dtypes:
            return self._check_fp8_dtype(dtype)

def get_device_interface(device: str) -> DeviceInterface:
    if device.type == 'cuda':
        return CUDADeviceInterface(device)
    return DeviceInterface(device)

```

- Instantiate test cases based on supported data types.
```diff
diff --git a/torch/testing/_internal/common_device_type.py b/torch/testing/_internal/common_device_type.py
index 834e0ed1007..f883afd3dea 100644
--- a/torch/testing/_internal/common_device_type.py
+++ b/torch/testing/_internal/common_device_type.py
@@ -488,8 +488,18 @@ class DeviceTypeTestBase(TestCase):
         # Parametrization decorators set the parametrize_fn attribute on the test.
         parametrize_fn = getattr(test, "parametrize_fn", default_parametrize_fn)

+        cls_device_type = (
+            cls.device_type
+            if cls.device_type != "privateuse1"
+            else torch._C._get_privateuse1_backend_name()
+        )
+
         # If one of the @dtypes* decorators is present, also parametrize over the dtypes set by it.
         dtypes = cls._get_dtypes(test)
+        device_valid_dtypes = get_device_interface(cls_device_type).supported_dtypes()
+        # Intersect the dtypes with the valid dtypes for the device type.
+        dtypes = list(set(dtypes).intersection(device_valid_dtypes))
+
         if dtypes is not None:

             def dtype_parametrize_fn(test, generic_cls, device_cls, dtypes=dtypes):
```

@malfet , does it make sense?

**Comment by EikanWang** [comment]
The device abstraction described here is intended specifically for the testing infrastructure. As such, I propose separating this test-focused device abstraction from the core torch runtime to maintain a clear separation of concerns. Nonetheless, this design is not meant to be exclusive: if certain APIs prove valuable to the core torch runtime, we should propose adding them to the core torch runtime.

**Comment by kundaMwiza** [comment]
@EikanWang Thank you for the proposal. I think [this proposal](https://github.com/pytorch/pytorch/issues/149734) could also be handled by this feature.
